File: coding/src/com/algos/topics/li10_tree/bt/revised/Bt_L102_BinaryTreeLevelOrderTraversal.java
package  com.algos.topics.li10_tree.bt.revised;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

import com.algos.topics.models.TreeNode;


public class Bt_L102_BinaryTreeLevelOrderTraversal {
    public static void main(String[] args) {
        Bt_L102_BinaryTreeLevelOrderTraversal obj = new Bt_L102_BinaryTreeLevelOrderTraversal();

        TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);

        System.out.println(obj.levelOrder(root));
    }

    // Problem: https://leetcode.com/problems/binary-tree-level-order-traversal/
    // Idea: Rely on queue size and process the queue elements until its size. Keep insert elements into same queue.
    // Time: O(n)
    // space: O(n)
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null)
            return ans;

        Queue<TreeNode> q = new LinkedList<>();
		q.add(root);
		while(!q.isEmpty()) {
            List<Integer> level = new ArrayList<>();
            int size = q.size();                // Take to a variable as q size changes
			for(int i=0; i<size; i++) {
				TreeNode curr = q.remove();
				level.add(curr.val);
                if(curr.left != null)
                    q.add(curr.left);
                if(curr.right != null)
                    q.add(curr.right);
			}

            ans.add(level);
		}
        return ans;
    }
}

File: dsa/topics/10_trees/bt/python/bt_l102__binary_tree_level_order_traversal.py
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/binary-tree-level-order-traversal/
# Idea: Use queue size to process nodes level by level. Add children to queue.
# Time: O(n)
# Space: O(n)

def level_order(root):
    ans = []
    if not root:
        return ans

    q = deque([root])

    while q:
        level = []
        size = len(q)  # Store current level size as queue size changes
        for _ in range(size):
            curr = q.popleft()
            level.append(curr.val)
            if curr.left:
                q.append(curr.left)
            if curr.right:
                q.append(curr.right)
        ans.append(level)

    return ans


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(level_order(root))  
# Expected: [[1], [3, 4], [2, 6], [5]]
